sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,370 @@
+    self.names = names or []
+    self.quality = quality
+    self.filters = filters or []
+    self.info = info or {}
+    self.calls = calls or []
+
+  def __eq__(self, other):
+    return (self.reference_name == other.reference_name and
+            self.start == other.start and
+            self.end == other.end and
+            self.reference_bases == other.reference_bases and
+            self.alternate_bases == other.alternate_bases and
+            self.names == other.names and
+            self.quality == other.quality and
+            self.filters == other.filters and
+            self.info == other.info and
+            self.calls == other.calls)
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.reference_name,
+                          self.start,
+                          self.end,
+                          self.reference_bases,
+                          self.alternate_bases,
+                          self.names,
+                          self.quality,
+                          self.filters,
+                          self.info,
+                          self.calls]])
+
+
+class VariantCall(object):
+  """A class to store info about a variant call.
+
+  A call represents the determination of genotype with respect to a particular
+  variant. It may include associated information such as quality and phasing.
+  """
+
+  def __init__(self, name=None, genotype=None, phaseset=None, info=None):
+    """Initialize the :class:`VariantCall` object.
+
+    Args:
+      name (str): The name of the call.
+      genotype (List[int]): The genotype of this variant call as specified by
+        the VCF schema. The values are either `0` representing the reference,
+        or a 1-based index into alternate bases. Ordering is only important if
+        `phaseset` is present. If a genotype is not called (that is, a `.` is
+        present in the GT string), -1 is used
+      phaseset (str): If this field is present, this variant call's genotype
+        ordering implies the phase of the bases and is consistent with any other
+        variant calls in the same reference sequence which have the same
+        phaseset value. If the genotype data was phased but no phase set was
+        specified, this field will be set to `*`.
+      info (dict): A map of additional variant call information. The key is
+        specified in the VCF record and the type of the value is specified by
+        the VCF header FORMAT.
+    """
+    self.name = name
+    self.genotype = genotype or []
+    self.phaseset = phaseset
+    self.info = info or {}
+
+  def __eq__(self, other):
+    return (self.name == other.name and
+            self.genotype == other.genotype and
+            self.phaseset == other.phaseset and
+            self.info == other.info)
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class _VcfSource(TextSource):
```

Comment:
Is it possible to do this without adding more inheritance to the Source hierarchy? The PTransform Style Guide heavily encourages usage of PTransform composition instead of inheritance: https://beam.apache.org/contribute/ptransform-style-guide/#injecting-user-specified-behavior ; even if you can't express this via PTransform composition, please try expressing it by using a TextSource as a member variable instead of inheriting from it - this will be significantly easier to maintain for the authors of TextSource (it was not designed to be inherited from)

Reply:
I changed this to use composition. To keep the functionality where `TextSource` can optionally read headers (as is required for `VcfSource`), I changed the `read_records` function to optionally take a `process_header_fn`.